### PR TITLE
Add HTML table operations

### PR DESCRIPTION
**Pull request by HenryL27**
Parse html tables into sycamore.data.Table objects.
Use that in HtmlPartitioner
Add TEDs implementation for Table objects.

Next thing is to squish the TEDs into the eval framework and run PubTabNet or something through it, but that's a different PR

### CODE REVIEW

File: lib/sycamore/sycamore/data/table.py [hunk 1]
```diff
@@ -143,6 +145,96 @@ def from_dict(cls, dict_obj: dict[str, Any]) -> "Table":
...
+                    )
+                )
+
+                cur_col += colspan
+
+            elif tag.name == "caption":
+                caption = tag.get_text()
+
+            else:
+                continue
```
Comment (baitsguy) [inline lib/sycamore/sycamore/data/table.py:219]: remove or maybe log if we shouldn't be here?

File: lib/sycamore/sycamore/data/table.py [hunk 2]
```diff
@@ -143,6 +145,96 @@ def from_dict(cls, dict_obj: dict[str, Any]) -> "Table":
...
+
+        # Traverse the tree of elements in a pre-order traversal.
+        for tag in root.find_all(recursive=True):
+            if tag.name == "tr":
+                cur_row += 1  # TODO: Should this be based on rowspan?
+                cur_col = 0
+
+            elif tag.name in {"td", "th"}:
+                # We allow a missing tr for the first row to handle the case when
+                # they have a thead.
```
Comment (baitsguy) [inline lib/sycamore/sycamore/data/table.py:195]: is it possible to have multiple theads? Can a single table contain 2 header rows

File: lib/sycamore/sycamore/data/table.py [hunk 3]
```diff
@@ -143,6 +145,96 @@ def from_dict(cls, dict_obj: dict[str, Any]) -> "Table":
...
+                continue
+
+        # Fix columns where rowspans should be inserted
+        for c in cells:
+            if len(c.rows) == 1:
+                continue
+            rows_to_increment = c.rows[1:]
+            increment_by = len(c.cols)
+            increment_after = c.cols[0]
+            # Yes this is quadratic. I'm assuming tables are smaller than LLMs.
```
Comment (baitsguy) [inline lib/sycamore/sycamore/data/table.py:228]: lol

File: lib/sycamore/sycamore/transforms/explode.py [hunk 4]
```diff
@@ -36,6 +37,8 @@ def explode(parent: Document) -> list[Document]:
             cur = Document(element.data)
             cur.doc_id = str(uuid.uuid4())
             cur.parent_id = parent.doc_id
+            if isinstance(element, TableElement):
```
Comment (baitsguy) [inline lib/sycamore/sycamore/transforms/explode.py:40]: why do we need special handling here?

File: lib/sycamore/sycamore/transforms/partition.py [hunk 5]
```diff
@@ -332,24 +332,9 @@ def partition(self, document: Document) -> Document:
                 if len(table.find_all("table")) > 0:
                     continue
 
-                table_element = TableElement()
```
Comment (baitsguy) [inline lib/sycamore/sycamore/transforms/partition.py:335]: nice

**Review by baitsguy — COMMENTED** [review]
Looks good, just couple small comments

File: lib/sycamore/sycamore/data/table.py (hunk 2 above)
Comment (HenryL27) [reply lib/sycamore/sycamore/data/table.py:195]: yep

File: lib/sycamore/sycamore/transforms/explode.py (hunk 4 above)
Comment (HenryL27) [reply lib/sycamore/sycamore/transforms/explode.py:40]: In a previous PR I changed the behavior of TableElement.text_representation to 
```
if overridden: overridden_text
else if table: table.to_csv()
else: text_representation
```
(and setting text_representation sets the overridden).

But just importing the underlying data essentially deletes this behavior when the table gets promoted to a document

**Review by baitsguy — APPROVED** [review]

**Pull request merged**